### DISCUSSION TITLE
Background Worker

### DISCUSSION THREAD

**Discussion by luigixiv** [ServiceStack/Discuss#51] [Q&A] [answered]
Hi,
I've a .net framework 4.8 server hosted in IIS which uses an "AppHostHttpListenerBase" instantiated inside the Global.asax of the application.
It loads dynamically the assemblies needed to run the various modules of the system using the base constructor of "AppHostHttpListenerBase" and then it calls the Init method on the host.
There is the possibility to get one of the loaded assembly and to call a method which will perform a background task which executes for the entire life of the server? (I mean until a restart or stop of IIS).
Or there is another way to achieve this behavior?

The client then will call an endpoint in the loaded service which will answer with data from that background task.
I need the task because the service must connect to an external resource and gets continuous data from that resource, then when a client wants those information will call the endpoint and gets those data.

I hope it is clear.
Thanks in advance.

Bye.

**Comment by mythz** [comment, marked as answer]
You can dynamically load an Assembly and call a method on App Startup that's run in a background thread which will run until it fails or is aborted.
 
But HttpListener Apps are typically run as a stand-alone Web Server, I've not heard of anyone running them inside IIS, which given IIS recycles App Domains I wouldn't think it's a good choice for long running Services unless it's configured to never recycle Apps.

You can run HttpListener Apps as a stand-alone Console App or run it inside a Windows Service as done in the [winservice-netfx](https://github.com/NetFrameworkTemplates/winservice-netfx) project template.

Whilst newer ASP .NET Core Apps can use [Worker Services](https://learn.microsoft.com/en-us/dotnet/core/extensions/workers) for long running services of which we have project templates for them on [ServiceStack's Start page](https://servicestack.net/start#workers).

**Reply by luigixiv** [reply]
Yes, you are right, with .Net Core everything is more simple, but for now I cannot migrate from .net framework to .net6.
Actually this server is like a windows service, but it is designed to be hosted inside IIS, because it exposes many endpoints to it's client and they communicate to each other through the classic http request/response flow (using a ServiceStack Json http client).

I think I will use a background thread created inside the Global.asax after initialization for the server part.
But what about the "ServerEvents" for the client side? It could be a good choice to add in the client the possibility to get notified by the server each time something changes in the Server? Where the changes are taken from that background worker.

Thanks.

**Reply by mythz** [reply]
Sure, clients can connect to the servers SSE /event-stream to get notified of updates, could long poll, periodic poll, use an intermediate MQ Broker for decoupling and added resiliency, or if clients are running their own HTTP servers the server could respond by calling a webhook callback. Lots of options.
